controllers/awsmachine_controller.go

```diff
@@ -202,13 +244,48 @@ func (r *AWSMachineReconciler) SetupWithManager(mgr ctrl.Manager, options contro
 			UpdateFunc: func(e event.UpdateEvent) bool {
 				oldCluster := e.ObjectOld.(*clusterv1.Cluster)
 				newCluster := e.ObjectNew.(*clusterv1.Cluster)
-				return oldCluster.Spec.Paused && !newCluster.Spec.Paused
+				log := r.Log.WithValues("predicateEvent", "update", "namespace", newCluster.Namespace, "cluster", newCluster.Name)
+
+				switch {
+				// never return true for a paused Cluster
+				case newCluster.Spec.Paused:
+					log.V(4).Info("Cluster is paused, will not reconcile associated AWSMachine.")
+					return false
+				// return true if Cluster.Status.InfrastructureReady has changed from false to true
+				case !oldCluster.Status.InfrastructureReady && newCluster.Status.InfrastructureReady:
+					log.V(4).Info("Cluster InfrastructureReady became ready, will reconcile associated AWSMachine.")
+					return true
```

Comment:
@ncdc I'm having trouble reproducing the delayed reconciliation on pivot, I've tried with the current version of this as well as replicating the same predicate logic from before with additional logging. That said, I am testing with the updates to controller-runtime and capi, but I wouldn't expect either of those to have a meaningful effect.

Even if it doesn't solve the delayed resync, I think it'll still be valuable to also trigger on InfrastructureReady changes here.

Reply:
Yeah, I've had a lot of trouble too. I think I've managed to see the delay twice in total.

The trigger on infraReady sgtm